Source/Core/VideoBackends/Vulkan/Renderer.cpp

```diff
@@ -737,22 +737,133 @@ bool Renderer::DrawFrameDump(const EFBRectangle& rc, u32 xfb_addr,
   return true;
 }
 
-void Renderer::DumpFrame(u64 ticks)
+void Renderer::StartFrameDumping()
+{
+  _assert_(!m_frame_dumping_active);
+
+  // Register fence callback so that we know when frames are ready to be written to the dump.
+  // This is done by clearing the fence pointer, so WriteFrameDumpFrame doesn't have to wait.
+  auto queued_callback = [](VkCommandBuffer, VkFence) {};
+  auto signaled_callback = std::bind(&Renderer::OnFrameDumpImageReady, this, std::placeholders::_1);
+
+  // We use the array pointer as a key here, that way if Renderer needed fence callbacks in
+  // the future it could be used without conflicting.
+  // We're not interested in when fences are submitted, so the first callback is a no-op.
+  g_command_buffer_mgr->AddFencePointCallback(
+      m_frame_dump_images.data(), std::move(queued_callback), std::move(signaled_callback));
+  m_frame_dumping_active = true;
+}
+
+void Renderer::EndFrameDumping()
 {
-  AVIDump::Frame state = AVIDump::FetchState(ticks);
-  DumpFrameData(reinterpret_cast<const u8*>(m_frame_dump_readback_texture->GetMapPointer()),
-                static_cast<int>(m_frame_dump_render_texture->GetWidth()),
-                static_cast<int>(m_frame_dump_render_texture->GetHeight()),
-                static_cast<int>(m_frame_dump_readback_texture->GetRowStride()), state);
+  _assert_(m_frame_dumping_active);
+
+  // Write any pending frames to the frame dump.
+  FlushFrameDump();
+
+  // Remove the fence callback that we registered earlier, one less function that needs to be
+  // called when preparing a command buffer.
+  g_command_buffer_mgr->RemoveFencePointCallback(m_frame_dump_images.data());
+  m_frame_dumping_active = false;
+}
+
+void Renderer::OnFrameDumpImageReady(VkFence fence)
+{
+  for (FrameDumpImage& frame : m_frame_dump_images)
+  {
+    // fence being a null handle means that we don't have to wait to re-use this image.
+    if (frame.fence == fence)
+      frame.fence = VK_NULL_HANDLE;
+  }
+}
+
+void Renderer::WriteFrameDumpImage(size_t index)
+{
+  FrameDumpImage& frame = m_frame_dump_images[index];
+  _assert_(frame.pending);
+
+  // Check fence has been signaled.
+  // The callback here should set fence to null.
+  if (frame.fence != VK_NULL_HANDLE)
+  {
+    g_command_buffer_mgr->WaitForFence(frame.fence);
+    _assert_(frame.fence == VK_NULL_HANDLE);
+  }
+
+  // Copy the now-populated image data to the output file.
+  AVIDump::Frame state = AVIDump::FetchState(frame.ticks);
   FinishFrameData();
+  DumpFrameData(reinterpret_cast<const u8*>(frame.readback_texture->GetMapPointer()),
+                static_cast<int>(frame.readback_texture->GetWidth()),
+                static_cast<int>(frame.readback_texture->GetHeight()),
+                static_cast<int>(frame.readback_texture->GetRowStride()), state);
+
+  frame.pending = false;
+}
+
+StagingTexture2D* Renderer::PrepareFrameDumpImage(size_t index, u32 width, u32 height, u64 ticks)
```

Comment:
This function accesses m_current_frame_dump_image and has an index parameter. This is a bit redundant.